sleap_io/model/camera.py

```diff
@@ -20,6 +81,181 @@ class CameraGroup:
+                M is the number of camera views and "..." is any number of dimensions
+                (including 0).
+            triangulation_func: Function to use for triangulation. The
+            triangulation_func should take the following arguments:
+                - points: Array of undistorted 2D points from each camera view of shape
+                    (M, N, 2) where M is the number of cameras and N is the number of
+                    points.
+                - projection_matrices: Array of (3, 4) projection matrices for each of
+                    the M cameras of shape (M, 3, 4) - note that points are undistorted.
+                and return the triangulated 3D points of shape (N, 3).
+            Default is vectorized DLT.
+
+
+        Raises:
+            ValueError: If points are not of shape (M, ..., 2).
+            ValueError: If number of cameras M do not match number of cameras in group.
+            ValueError: If number of points returned by triangulation function does not
+                match number of points in input.
+
+        Returns:
+            Triangulated 3D points of shape (..., 3) where "..." is any number of
+            dimensions and matches the "..." dimensions of the input points array.
+        """
+        # Validate points in
+        points_shape = points.shape
+        try:
+            n_cameras = points_shape[0]
+            if n_cameras != len(self.cameras):
+                raise ValueError
+            if 2 != points.shape[-1]:
+                raise ValueError
+            if len(points_shape) != 3:
+                points = points.reshape(n_cameras, -1, 2)
+        except Exception as e:
+            raise ValueError(
+                "Expected points to be an array of 2D points from each camera view of "
+                f"shape (M, ..., 2) where M = {len(self.cameras)} and '...' is any "
+                f"number of dimensions, but received shape {points_shape}.\n\n{e}"
+            )
+        n_points = points.shape[1]
+
+        # Undistort points
+        points = points.astype("float64")  # Ensure float64 for opencv undistort
+        for cam_idx, camera in enumerate(self.cameras):
+            cam_points = camera.undistort_points(points[cam_idx])
+            points[cam_idx] = cam_points
+
+        # Since points are undistorted, use extrinsic matrices as projection matrices
+        projection_matrices = np.array(
+            [camera.extrinsic_matrix[:3] for camera in self.cameras]
+        )
+
+        # Triangulate points using provided function
+        points_3d = triangulation_func(points, projection_matrices)
+        n_points_returned = points_3d.shape[0]
+        if n_points_returned != n_points:
+            raise ValueError(
+                f"Expected triangulation function to return {n_points} 3D points, but "
+                f"received {n_points_returned} 3D points."
+            )
+
+        return points_3d.reshape(*points_shape[1:-1], 3)
```

Comment:
Added some logic for preserving dtype.

```suggestion
        # Undistort points
        points_dtype = points.dtype
        points = points.astype("float64")  # Ensure float64 for opencv undistort
        for cam_idx, camera in enumerate(self.cameras):
            cam_points = camera.undistort_points(points[cam_idx])
            points[cam_idx] = cam_points

        # Since points are undistorted, use extrinsic matrices as projection matrices
        projection_matrices = np.array(
            [camera.extrinsic_matrix[:3] for camera in self.cameras]
        )

        # Triangulate points using provided function
        points_3d = triangulation_func(points, projection_matrices)
        n_points_returned = points_3d.shape[0]
        if n_points_returned != n_points:
            raise ValueError(
                f"Expected triangulation function to return {n_points} 3D points, but "
                f"received {n_points_returned} 3D points."
            )

        # Reshape to (N, 3) and cast to the original dtype.
        points_3d = points_3d.reshape(*points_shape[1:-1], 3).astype(points_dtype)
        return points_3d
```